### PR TITLE
Allow using and registering skyboxes in code

### PR DESCRIPTION
**Pull request by BoogieMonster1O1**
This pr basically does three things
- Allows adding permanent skyboxes that are stored in a list that is not cleared during a resource reload.
- Replace suppliers that were used to get codecs and types with `SkyboxType` and a registry.
- Replace legacy deserialization code inside skyboxes with separate deserializers. Also uses a registry.

I've also added fabric api as a dependency instead of just resource loader so that I can use registry sync's registry registering hooks

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -31,8 +31,8 @@ dependencies {
     modImplementation "net.fabricmc:fabric-loader:${project.loader_version}"
 
     // fabric api resource loader
-    modCompile "net.fabricmc.fabric-api:fabric-resource-loader-v0:${project.fabric_resource_loader_version}"
-    include "net.fabricmc.fabric-api:fabric-resource-loader-v0:${project.fabric_resource_loader_version}"
+    modCompile "net.fabricmc.fabric-api:fabric-api:${project.fabric_version}"
+    include "net.fabricmc.fabric-api:fabric-api:${project.fabric_version}"
```
Comment (AMereBagatelle) [inline build.gradle]: Yeah, no.  I personally don't like the idea of including all of FAPI.  What other parts of it do you use?

**Review by AMereBagatelle — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (FlashyReese) [reply build.gradle]: It's implementing only

Comment (BoogieMonster1O1) [reply build.gradle]: to answer your question: resource loader and registry sync
jijing registry sync is concern and its something nobody ever does

**Comment by BoogieMonster1O1** [comment]
The Test mod works. Now ready for review

File: src/main/java/io/github/amerebagatelle/fabricskyboxes/SkyboxManager.java [hunk 2]
```diff
@@ -22,49 +22,36 @@
...
-    }
-
-    public static List<Supplier<? extends AbstractSkybox>> getSkyboxTypes() {
-        return SKYBOX_TYPES;
-    }
-
-    private static final ArrayList<AbstractSkybox> skyboxes = new ArrayList<>();
+    private final Predicate<? super AbstractSkybox> renderPredicate = (skybox) -> !this.activeSkyboxes.contains(skybox) && skybox.alpha >= 0.1;
+    private final ArrayList<AbstractSkybox> skyboxes = new ArrayList<>();
+    private final ArrayList<AbstractSkybox> permanentSkyboxes = new ArrayList<>();
```
Comment (AMereBagatelle) [inline src/main/java/io/github/amerebagatelle/fabricskyboxes/SkyboxManager.java:35]: I personally require a little more description on all of this, seems a little unclear what permanentSkyboxes actually does for us

File: src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import com.mojang.serialization.Codec;
+import io.github.amerebagatelle.fabricskyboxes.skyboxes.AbstractSkybox;
+import io.github.amerebagatelle.fabricskyboxes.skyboxes.MonoColorSkybox;
+import io.github.amerebagatelle.fabricskyboxes.skyboxes.SkyboxType;
+import io.github.amerebagatelle.fabricskyboxes.util.object.Conditions;
+import io.github.amerebagatelle.fabricskyboxes.util.object.Decorations;
+import io.github.amerebagatelle.fabricskyboxes.util.object.DefaultProperties;
+import io.github.amerebagatelle.fabricskyboxes.util.object.RGBA;
+
+public class AnEntirelyHardcodedSkybox extends MonoColorSkybox {
```
Comment (AMereBagatelle) [inline src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java]: Change the name, something like "HardcodedTestSkybox" would be better.

File: src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package io.github.amerebagatelle.fabricskyboxes;
+
+import io.github.amerebagatelle.fabricskyboxes.skyboxes.SkyboxType;
+
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.api.ClientModInitializer;
+
+public class Init implements ClientModInitializer {
```
Comment (AMereBagatelle) [inline src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java]: Change name to something like "FSTestInit"

**Review by AMereBagatelle — CHANGES_REQUESTED** [review]
Looks good in general, still have to test it locally but here's the things I found in code

File: src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java (hunk 3 above)
Comment (BoogieMonster1O1) [reply src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java]: this is the testmod

File: src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java (hunk 4 above)
Comment (BoogieMonster1O1) [reply src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java]: :eyes:
sure

File: src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java (hunk 3 above)
Comment (AMereBagatelle) [reply src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java]: I know, and that's why.  I don't feel like `AnEntirelyHardcodedSkybox` really is good code style naming

File: src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java (hunk 4 above)
Comment (BoogieMonster1O1) [reply src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java]: but its the testmod

Comment (AMereBagatelle) [reply src/testmod/java/io/github/amerebagatelle/fabricskyboxes/Init.java]: I know it's picky... it's a personal code style thing :P

File: src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java (hunk 3 above)
Comment (BoogieMonster1O1) [reply src/testmod/java/io/github/amerebagatelle/fabricskyboxes/AnEntirelyHardcodedSkybox.java]: the testmod is mostly implementation detail, but sure ill rename it

**Comment by BoogieMonster1O1** [comment]
Renamed testmod classes and added javadoc. 

I also annotated some of the methods in `SkyboxManager` to specify that they're not part of the public api (they should also probably be moved to a separate class (`SkyboxManagerImpl`?) in a later version)

**Comment by AMereBagatelle** [comment]
I was considering doing something of the sort to SkyboxManager, also I think this is now ready for merge

File: docs/api.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Api Documentation
+### Permanent skyboxes
+FabricSkyBoxes provides an api that allows you to register skyboxes in code. 
+These skyboxes will never be cleared after a resource reload and should often only be used to provide custom sky rendering to your custom dimension.
```
Comment (AMereBagatelle) [inline docs/api.md]: Or biome :D

Comment (BoogieMonster1O1) [reply docs/api.md]: fixed

**Pull request merged**